src/socket/icmp.rs

```diff
@@ -0,0 +1,569 @@
+use core::cmp;
+use managed::Managed;
+
+use {Error, Result};
+use phy::{ChecksumCapabilities, DeviceCapabilities};
+use socket::{Socket, SocketHandle};
+use storage::{Resettable, RingBuffer};
+use wire::{IpAddress, IpEndpoint, IpProtocol, IpRepr};
+use wire::{Ipv4Address, Ipv4Repr};
+use wire::{Icmpv4Packet, Icmpv4Repr};
+use wire::{UdpPacket, UdpRepr};
+
+/// Type of endpoint to bind the ICMP socket to. See [IcmpSocket::bind] for
+/// more details.
+///
+/// [IcmpSocket::bind]: struct.IcmpSocket.html#method.bind
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+pub enum Endpoint {
+    Identifier(u16),
+    IpEndpoint(IpEndpoint),
```

Comment:
`Endpoint::IpEndpoint`? I suggest just `Ip`

Reply:
What about `Udp`? Since you're binding to ICMP `DstUnreachables` for UDP packets from the given port.

Reply:
I'm okay with `Ip` though.. just trying to think about how to make this as readable as possible